### ISSUE TITLE
CHROME HEADLESS support -- in progress

### ISSUE THREAD

**Issue by garris** [garris/BackstopJS#454]
I think if BackstopJS is to remain relevant we need to move to headless chrome as soon as possible.   I can see a clear path to this by swapping out Casper for this https://github.com/OnetapInc/chromy

Backstoppers -- what do we think???

**Comment by garris** [comment]
^^^ @klausbayrhammer, @jtuds, @svwu, @mirzazeyrek, @primajin, @stevecfischer

**Comment by garris** [comment]
^^^ @onigoetz

**Comment by Primajin** [comment]
Absolutely! We need to do it right now or never to get the word out there to our user base. We also need to improve the documentation further because I found many of the [issues](https://github.com/garris/BackstopJS/issues) are related to unclear docs. For example many people confuse CLI and CI and so on. I'll jump in this week and try to come up with a more "eli5" documentation and play around with chrome headless.

**Comment by onigoetz** [comment]
I agree with this, but I would also encourage to support both implementations for a time. 
Phantomjs currently has the advantage of being packaged as an npm module and easy to install.

Chrome currently has to be installed separately, and it can go against the ease of use of backstop.

What do you think ?

**Comment by garris** [comment]
Yes -- totally agree.  Maybe start a v3 branch or even fork and rename backstop-chromy and maintain feature parity for a good long time.   

It's also totally possible the Casper guys will end up implementing headless chrome -- if they do then it's totally possible that we might drop the chromy implementation and go back to the Casper implementation.

**Comment by garris** [comment]
@Primajin  totally agree -- thanks for helping with that!

**Comment by mirzazeyrek** [comment]
@garris I was also thinking about this. If we start an open source project implementing headless chrome into CasperJS, I am sure there will be a lot of contributions.

CasperJS maintainers are very nice people and they are always open for such contributions.

On the other hand, implementing Chromy looks more easier but how this will effect backwards compatibility with CasperJS scripts ?

**Comment by svwu** [comment]
I'm totaly up for chromy!

It seems to be not too hard to change to it.
Would also go for WebDriver, but that's a far away vision, isn't it?

**Comment by klausbayrhammer** [comment]
I really like the idea to support headless chrome. Chromy looks promising although it's using async/await quite heavily so we would either have to transpile it or drop support for node<7. But when we start things of a new branch then moving to node7+ could also be an option

**Comment by garris** [comment]
@mirzazeyrek yes -- the Casper guys are a great bunch and they are also already looking into headless.   I think they will eventually support it -- but I think it could take a very long time.  My concern is around relevance.  1) There is a lot of community interest in headless chrome.  2) phantom is frustrating because support for new features is weak.  So I think if given the chance, the community would gladly move to a solution where there is no requirement for special shims or polyfills or weird implementations just for testing.

The upside of a chromy version is that chromy is written in JS which makes contributions to that project much easier for our community.  The bar for contributing to Casper is much higher -- so that makes a project like chromy much more compatible with backstop.

You do bring up a very good point about the onBefore, onReady scripts.   But I think the goal would be to continue to also offer BOTH versions for the foreseeable future.   That said, at LinkedIn we are using a lot of helper scripts to set cookies and manipulate the UI, but, even though we have many scripts, its only a handful of methods that we are actually using.  So, potentially, stubbing the Casper object with a handful of "most common" calls would enable most scripts to work without the need for refactoring. 

If that's the case we could potentially decouple from Casper all together.   OTOH -- if they catch up and eventually release a headless option we could make a decision about dropping support for one or the other.   

Phantom was a groundbreaking app (much respect!) -- but it will inevitably be go the way of YUI and mootools 😛.   So it's good to be having this conversation now.

**Comment by garris** [comment]
@svwu Yes, I like the webdriver option a lot -- but it appears that a chromy implementation would be a relatively easy switch.

**Comment by garris** [comment]
@klausbayrhammer Good point.  I am totally open to the following (in current order of preference)
- supporting just 7+ (simplest)
- polyfilling for a few methods (more complex but not a big deal)
- setting up transpiler  (more trouble than it's worth?  I currently like that current codebase needs no build.  But then, I think some devs _would_ like this option -- so I am open to it if others really like this approach)

**Comment by garris** [comment]
Alrighty then -- The people have spoken!  

https://developers.google.com/web/updates/2017/04/headless-chrome

https://github.com/OnetapInc/chromy

- I will start a branch: backstop-chromy

- if it's possible -- maybe we only need to swap out the capture module?  If so, this could make maintaining feature parity easier.

**Comment by davidjgoss** [comment]
@garris that sounds fantastic. I've oddly enough only come to this in the last couple of days and found it much nicer to use than the alternatives. The only problem is PhantomJS' non-support of web fonts, and this would certainly solve that.

Hit me up if you need any testing done on Windows!

**Comment by garris** [comment]
@davidjgoss   Thanks, will do!

**Comment by klausbayrhammer** [comment]
We could also try to slim down the capture-module to the essentials only (so e.g. removing config default values, computation of scenario names, ..). This would make the api we need to implement for headless chrome much clearer.

**Comment by garris** [comment]
+1

**Comment by garris** [comment]
Update.   I am about half-way to having chromy integrated.

https://github.com/garris/BackstopJS/tree/backstop-chromy-working

Created a file called `runChromy.js` -- this is called by createBitmaps (which currently only calls `runCasper.js`)

`runChromy.js` follows the same format as genConfig.js -- but since it runs in node-world it can be broken down into separate files eventually.   You can see the basic existing structure can be used by applying chromy syntax where casper currently is.

It's not producing output yet.

**Comment by garris** [comment]
One more thing.  This version will be able to capture screen shots in parallel -- which could be a huge perf win.

**Comment by garris** [comment]
Good news.  Simple integration is completed!

To use, set 
```
engine: `chromy`
```


You can Install here... 
```
npm install backstopjs@beta
```

branch is here... 
```
https://github.com/garris/BackstopJS/tree/backstop-chromy
```

Please note, this is only a simple implementation -- this will capture selectors and run a comparison. Reference, test, approve, genConfig are all working but some of the advanced features have not been added back yet.

Here is a list of config features not yet wired up...
- `debug` (logging errors, resources, markup to CLI)
- `readyEvent`
- `onBeforeScript`
- Auth User via CLI
- `onReadyScript`
- `hideSelectors`
- `removeSelectors`
- `selectorExpansion`
- capturing the whole document
- also, Travis seems very pissed off at me for making these changes -- so that needs to be fixed.

There are some expected differences between chromy and casper screenshots as well.  These will also need to be dealt with.  I will post those issues in chromy issues.

I am more than happy if anyone wants to work on adding any of the above features back.  We might need to break those into different files to make merging less challenging.

Cheers. 😸

**Comment by NimaSoroush** [comment]
Hi @garris,

My first attempt to run this with chromy didn't succeed. Should be node babel issue!

```x:simpleReactApp daphnebogosian$ npm run build

> first-webapp-react@1.0.0 build /Users/x/components/y/node_modules/backstopjs/examples/simpleReactApp
> NODE_ENV=production browserify -t [ reactify --es6 ] main.js | uglifyjs > compiled.js && backstop test

/Users/x/components/y/node_modules/chromy/src/index.js:73
  async start (startingUrl = null) {
        ^^^^^
SyntaxError: Unexpected identifier
    at Object.exports.runInThisContext (vm.js:76:16)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/x/components/y/node_modules/backstopjs/core/util/runChromy.js:1:78)
    at Module._compile (module.js:570:32)
```
should this be fixed by default?

**Comment by garris** [comment]
I am developing using node version 8.   It works with that version.  Please go ahead and add a transpile script issue to chromy if you'd like to use it with earlier versions.

**Comment by Primajin** [comment]
Cool! So I also played around a little bit, here are my findings:

It spawned a couple of real browser windows, I thought we are going for `chrome --headless` 🤔 ?
![screen shot 2017-06-26 at 17 13 35](https://user-images.githubusercontent.com/1742115/27546393-eec29d80-5a92-11e7-9df0-c4a1db121ca3.png)

I got `Failed to execute 'querySelector' on 'Document': '[class^=&quot;preview&quot;] button' is not a valid selector.`
for having `"[class^=\"preview\"] button"` as a selector, that was valid before.
Using `"[class^='preview'] button"` gives me `[class^=&#x27;preview&#x27;] button` as an error, so this seems to be a problem with (double) encoding. `[class^=preview] button` works.

Now finally the buttons get rendered how they should!
![screen shot 2017-06-26 at 17 28 38](https://user-images.githubusercontent.com/1742115/27547046-f25f1bd8-5a94-11e7-8d43-ab58968c2e83.png)

**Comment by garris** [comment]
@primajin thanks for testing!   

1) yes -- the intention is for chrome headless -- there is a config which allows for enable/disable of the window (see runChromy.js line 6).   I think this should be made available in the user config for troubleshooting.


2) it is possible that chrome won't close if there is a chromy error (like your selector issue)  otherwise -- with the latest version -- chromy is closing as expected.  Please make sure you're up to date and do let me know if running a default config still creates this error.

3) I just want to make sure I understand your comment regarding double encoding.  It looks like you need to specify the selector without inner quotes -- which to me is expected behavior as far as chrome is concerned.    Am I understanding this correctly or is this an issue and I am I missing your point?

**Comment by Primajin** [comment]
2. I checked out the beta an hour ago, so I think I got the latest version. 
3. Well actually the correct CSS selector is **with** quotes so you can specify if you want the selector to be case sensitive or not: https://developer.mozilla.org/pt-BR/docs/Web/CSS/Attribute_selectors (check line 30) but apart from that chrome is smart enough to guess what is meant, if you leave them out.

**Comment by garris** [comment]
Ok -- thanks for the clarification.  Sounds like quotes (which are otherwise legal) cause an error.   Do you know if this an issue with chromy or backstop?

**Comment by Primajin** [comment]
It worked before without problem with backstop and phantom but I'll dig into it tomorrow morning (CEST) to see if it is a chromy issue or came with the beta branch.

**Comment by garris** [comment]
That's a big help -- thanks.   

I spoke a little bit with the chromy project owner.  He is a very cool guy 😎.    If you discover it is a chromy issue you can go ahead and file it on that repo (and reference it here)

Cheers!

**Comment by dotneet** [comment]
@garris Thank you for introducing me!

@Primajin I'm owner of chromy. The css selector problem you found was chromy's issue.
I've fixed that problem and published new version. If it is reproduced yet after update chromy, please open an issue in chromy repo. Thank you for reporting.

**Comment by Primajin** [comment]
Great @dotneet , I'm currently setting up a docker image to run backstop via chromy in our CI 🎉 .
@garris runChromy.js line 6 did the trick. Maybe we can set it to false when report is set to CI and true when set to browser? Or we introduce a new backstop.js config property?

Otherwise it's hard for me to test it right now, since it lives inside node_modules.

**Comment by garris** [comment]
Great!  And yes, this window feature should always default to false and can be set to true by user in backstop config.

**Comment by Primajin** [comment]
@garris you said "capturing the whole document" is not yet implemented -
I read [here](https://developers.google.com/web/updates/2017/04/headless-chrome#screenshots) that headless chrome now directly has a screenshot feature. Maybe that `--screenshot` flag could come in handy there.

**Comment by garris** [comment]
@Primajin sorry -- completed that one yesterday https://github.com/garris/BackstopJS/commit/5da367b5d57b871bc0df1a3c8ee87dd0f2bc209b

**Comment by garris** [comment]
I am currently attempting to get scripts to work.

**Comment by garris** [comment]
onBefore scripts are now working.
backstopjs@beta

**Comment by Primajin** [comment]
cool! Since we are now on chromy in @beta - can we remove casper and phantom from the package.json? I am experiencing a very long build sometimes, because it tries to get all of the dependencies, but most of them are now not needed anymore because of chromy 🎉

**Comment by Primajin** [comment]
Also @garris the [recent change](https://github.com/garris/BackstopJS/commit/ea6bca828283b9f90de8f05be0543c7ff5538b64#diff-e3204928aaa5920f0162f7c1d91a486fR236) with adding the `.innerText;` fails my builds 😉

**Comment by garris** [comment]
@Primajin Thanks for actively testing on this branch!   At the moment I think we want to still support casper/phantom/slimer -- there are (estimated) thousands of users still relying on that implementation.   

Also,  I think NPM should cache most of the dependencies so that subsequent builds don't take quite as long?  But maybe this is not true for your use case?

**Comment by garris** [comment]
Apologies for the .innerText reference.  I will fix this.

**Comment by Primajin** [comment]
@garris Well I just thought as the headline states "swap out Casper for chromy" that those aren't needed anymore :wink:
And don't worry about the reference: It's not called a beta branch for no reason :wink:

**Comment by garris** [comment]
Oh yeah -- thank you for pointing this out!   When I first created this issue I _was_ considering a swap.   But I got some feedback that removing the casper would be disruptive to a lot of users -- so it seems the best route is to just add chrome at this point.   I will change the title -- I don't want to confuse others!

**Comment by garris** [comment]
Issues with onBefore and onReady scrips have been resolved.   Also that .innnerText has been fixed.

**Comment by Primajin** [comment]
Allright I retest, somehow my pngs are corrupted, but maybe it's a different issue 🤔 

![screen shot 2017-06-30 at 17 36 06](https://user-images.githubusercontent.com/1742115/27743017-f64b64ee-5dba-11e7-8b65-50731c65bef3.png)

**Comment by garris** [comment]
Hmm,  I haven't seen that.  Are you using the default config?   Checking for isVisible is not implemented yet.

**Comment by Primajin** [comment]
No, actually it's the same config as [five days ago](https://github.com/garris/BackstopJS/issues/454#issuecomment-311094059) - just a newer checkout of the npm module.

**Comment by garris** [comment]
Just curious what happens if you create a fresh directory, do `backstop genConfig` then `backstop test`.   Will that work?

**Comment by Primajin** [comment]
Hmm seem to resolved itself - strange. But furthermore I found that there are some weird files getting created in my project root folder: `/capture/` `/v8-compile-cache-0/` and a json file with a gibberish name. 🤔  Any ideas on that?

**Comment by garris** [comment]
+1 on the self-healing code! 😄  Glad that worked out.

Hmm I don't know anything about those files you mentioned.  My guess is that it is something generated by chrome and implemented in chromy.   See...  https://www.npmjs.com/package/v8-compile-cache

**Comment by garris** [comment]
closed by #479
